paymentsheet/src/main/java/com/stripe/android/paymentsheet/IntentConfirmationHandler.kt

```diff
@@ -0,0 +1,338 @@
+                    onFailure(
+                        cause = nextStep.cause,
+                        message = nextStep.message,
+                        type = ErrorType.NextStep,
+                    )
+                }
+                is IntentConfirmationInterceptor.NextStep.Complete -> {
+                    onPaymentResult(InternalPaymentResult.Completed(intent))
+                }
+            }
+        }
+    }
+
+    /**
+     * Waits for an intent result to be returned following a call to start an intent confirmation process through the
+     * [start] method. If no such call has been made, will return null.
+     *
+     * @return result of intent confirmation process or null if not started.
+     */
+    suspend fun awaitIntentResult(): Result? {
+        return completableResult?.await()
+    }
+
+    private fun handleNextAction(
+        clientSecret: String,
+        stripeIntent: StripeIntent,
+    ) = withPaymentLauncher { launcher ->
+        /*
+         * In case of process death, we should store that we waiting for a payment result to return from a
+         * payment confirmation activity
+         */
+        storeIsAwaitingForPaymentResult()
+
+        when (stripeIntent) {
+            is PaymentIntent -> {
+                launcher.handleNextActionForPaymentIntent(clientSecret)
+            }
+            is SetupIntent -> {
+                launcher.handleNextActionForSetupIntent(clientSecret)
+            }
+        }
+    }
+
+    private fun confirmStripeIntent(
+        confirmStripeIntentParams: ConfirmStripeIntentParams
+    ) = withPaymentLauncher { launcher ->
+        /*
+         * In case of process death, we should store that we waiting for a payment result to return from a
+         * payment confirmation activity
+         */
+        storeIsAwaitingForPaymentResult()
+
+        when (confirmStripeIntentParams) {
+            is ConfirmPaymentIntentParams -> {
+                launcher.confirm(confirmStripeIntentParams)
+            }
+            is ConfirmSetupIntentParams -> {
+                launcher.confirm(confirmStripeIntentParams)
+            }
+        }
+    }
+
+    private fun onPaymentResult(result: InternalPaymentResult) {
+        val intentResult = when (result) {
+            is InternalPaymentResult.Completed -> Result.Succeeded(
+                intent = result.intent,
+                deferredIntentConfirmationType = deferredIntentConfirmationType
+            )
+            is InternalPaymentResult.Failed -> Result.Failed(
+                cause = result.throwable,
+                message = result.throwable.stripeErrorMessage(),
+                type = ErrorType.Payment,
+            )
+            is InternalPaymentResult.Canceled -> Result.Canceled
+        }
+
+        deferredIntentConfirmationType = null
+
+        completableResult?.complete(intentResult)
+
+        removeIsAwaitingForPaymentResult()
+    }
+
+    private fun onFailure(
```

Comment:
Is there other cleanup we need to do here? Looks like we do a few other things in the success case a few lines above.